scipy/optimize/_zeros_py.py

```diff
@@ -1388,3 +1388,278 @@ def toms748(f, a, b, args=(), k=1,
+        Advances in Engineering Software, 28(3), 145-149.
+        https://doi.org/10.1016/s0965-9978(96)00051-8
+
+    See Also
+    --------
+    brentq, brenth, ridder, bisect, newton
+
+    Examples
+    --------
+    >>> from scipy import optimize
+    >>> def f(x,a):
+    ...     return a-x*x
+
+    >>> y = optimize._chandrupatla(f,0,3,args=(7.0,))
+    >>> y.x
+    2.6457513110645907
+
+    >>> x0 = np.zeros((2,1,1))
+    >>> x1 = np.full((1,2,1), 3)
+    >>> k = np.array([[[1, 2]]])
+    >>> res = optimize._chandrupatla(f,x0=x0,x1=x1,
+                              args=(k,),
+                              maxiter=50)
+    >>> res.x
+    [[[ 1.000e+00  1.414e+00]
+         [ 1.000e+00  1.414e+00]]
+
+        [[ 1.000e+00  1.414e+00]
+         [ 1.000e+00  1.414e+00]]]
+
+    permission from Jason Sachs (@jason-s):
+        https://github.com/scipy/scipy/issues/7242#issuecomment-1314178133
+    """
+
+    # adapted from an earlier implementation of Jason Sachs'
+    # as written in https://www.embeddedrelated.com/showarticle/855.php
+    # which in turn is based on Chandrupatla's algorithm as described in
+    # Scherer https://books.google.com/books?id=cC-8BAAAQBAJ&pg=PA95
+
+    # Initialization
+    b = x0
+    a = c = x1
+    fb = f(b, *args)
+    fa = fc = f(a, *args)
+    t = 0.5
+    iterations = 0
+
+    a, b, fa, fb = np.broadcast_arrays(a, b, fa, fb)
+    intermediate_shape = a.shape
+    a, b, fa, fb = np.atleast_1d(a, b, fa, fb)
+    shape = a.shape
+
+    iterations = np.zeros_like(a)
+
+    # flag to check state of convergence
+    flag = np.full(shape, _EINPROGRESS)
+
+    # Initialize an array of False,
+    # determines whether we should do inverse quadratic interpolation
+    iqi = np.zeros(shape, dtype=bool)
+
+    funcalls = np.full(np.shape(flag), 0)
+    terminate = False
+
+    while iterations.any() < maxiter:
+        # use t to linearly interpolate between a and b,
+        # and evaluate this function as our newest estimate xt
+        xt = a + t * (b - a)
+        ft = f(xt, *args)
+        funcalls += 1
+
+        # update our history of the last few points so that
+        # - a is the newest estimate (we're going to update it from xt)
+        # - c and b get the preceding two estimates
+        # - a and b maintain opposite signs for f(a) and f(b)
+        samesign = np.sign(ft) == np.sign(fa)
+        c = np.choose(samesign, [b, a])
+        b = np.choose(samesign, [a, b])
+        fc = np.choose(samesign, [fb, fa])
+        fb = np.choose(samesign, [fa, fb])
+        a = xt
+        fa = ft
+
+        # set r so that f(r) is the minimum magnitude of f(a) and f(b)
+        fa_is_smaller = np.abs(fa) < np.abs(fb)
+        r = np.choose(fa_is_smaller, [b, a])
+        fm = np.choose(fa_is_smaller, [fb, fa])
+
+        tol = 2 * xatol * np.abs(r) + xrtol
+        tlim = tol / np.abs(b - c)
+        terminate = np.logical_or(terminate,
+                                  np.logical_or(fm == 0, tlim > 0.5))
```

Comment:
One of the problems I'm having integrating this method with the tests is that this termination condition is different from that of all the other methods. Just so we can can get tests to pass, can you add an argument to `_chandrupatla` that makes it easy to switch between the termination condition here and the termination conditions used in other solvers (e.g. `bisect`)? This would only be used for testing purposes.

Reply:
Sure, I'll try to implement that.

Reply:
Something like this? I'm looking through `bisect.c` and `brenth.c` and as far as I can tell their termination conditions are somewhat like the below.

```python 
        tol = 2 * xatol * np.abs(r) + xrtol
        tlim = tol / np.abs(b - c)
        if test_switch:
            delta = (xatol + xrtol * abs(r))
            terminate = np.logical_or(terminate,
                                      np.logical_or(fm == 0,
                                                    abs(b - a) < delta))
        else:
            terminate = np.logical_or(terminate,
                                      np.logical_or(fm == 0, tlim > 0.5))

```

Reply:
Something like that, although
* I think `xatol` and `xrtol` appear to be mixed up in the calculation of `tol`.
* `tol` and `tlim` only need to be calculated in the `else` case.

In the original paper:
![image](https://user-images.githubusercontent.com/6570539/224510987-76b33739-aed6-44fb-85c5-ffe935cb7274.png)

Please also check whether it is correct to calculate `tlim` with `b` and `c`. My guess would have been `b` and `a`.

Reply:
Ah I didn't realize that was specified in the paper itself, thanks for pointing that out. I'll read through that section and make sure it's `b` and `a`. Though yes it does look like those are swapped in `tol`, will fix that as well.

Reply:
Again not an expert here, but it seems like this is saying that `[a,b]` will bracket the root, so I'll proceed with that in the calculation of tlim unless you disagree. 
![image](https://user-images.githubusercontent.com/42592742/224554674-7850f83d-abc4-4348-99ba-06a2981210c4.png)

Also, seems like we need `tlim` and `tol` for:

```python
# limit to the range (tlim, 1-tlim)
t = np.minimum(1 - tlim, np.maximum(tlim,t))
```

unless that is also somehow related to the termination condition that I'm not understanding.

Reply:
Ah okay, sounds great, thanks!